XMonad/Layout/SubLayouts.hs

```diff
@@ -442,20 +429,48 @@ updateWs = windowsMaybe . updateWs'
 
 updateWs' :: Groups Window -> WindowSet -> Maybe WindowSet
 updateWs' gs ws = do
-    f <- W.peek ws
-    let w = W.index ws
-        nes = concatMap W.integrate $ mapMaybe (flip M.lookup gs) w
-        ws' = W.focusWindow f $ foldr W.insertUp (foldr W.delete' ws nes) nes
-    guard $ W.index ws' /= W.index ws
-    return ws'
+    w <- W.stack . W.workspace . W.current $ ws
+    let w' = flattenGroupStack . toGroupStack gs $ w
+    guard $ w /= w'
+    pure $ W.modify' (const w') ws
+
+-- | Flatten a stack of stacks.
+flattenGroupStack :: GroupStack a -> W.Stack a
+flattenGroupStack (W.Stack (W.Stack f lf rf) ls rs) =
+    let l = lf ++ concatMap (reverse . W.integrate) ls
+        r = rf ++ concatMap W.integrate rs
+    in W.Stack f l r
+
+-- | Extract Groups from a stack of stacks.
+fromGroupStack :: (Ord a) => GroupStack a -> Groups a
+fromGroupStack = M.fromList . map (W.focus &&& id) . W.integrate
+
+-- | Arrange a stack of windows into a stack of stacks, according to (possibly
+-- outdated) Groups.
+toGroupStack :: (Ord a) => Groups a -> W.Stack a -> GroupStack a
+toGroupStack gs st@(W.Stack f ls rs) =
+    W.Stack (let Just f' = lu f in f') (mapMaybe lu ls) (mapMaybe lu rs)
+  where
+    wset = S.fromList (W.integrate st)
+    dead = W.filter (`S.member` wset) -- drop dead windows or entire groups
+    refocus s = focusWindow' f s <|> pure s -- sync focus into groups
+    gs' = mapGroups (refocus <=< dead) gs
+    gset = S.fromList . concatMap W.integrate . M.elems $ gs'
+    -- after refocus, f is either the focused window of some group, or not in
+    -- gs' at all, so `lu f` is never Nothing
+    lu w = if w `S.member` gset
+        then w `M.lookup` gs'
+        else Just (W.Stack w [] [])
```

Comment:
Wouldn't it be better to just `lu w = fromJust $ ...` here?  If this invariant is ever not satisfied, the current code will also blow up when doing `let Just f' = lu f in f'` above, so it wouldn't be any less safe and it would simplify the construction of the new stack above

Reply:
No, the invariant only holds for `f` (the focused window), not for all windows.
I don't like the `let Just f' = …` either but I don't think the invariant can be captured by Haskell's type system. :-/

Reply:
But I can surely improve the comment. :-)